src/main/java/com/thebois/models/inventory/AbstractInventory.java

```diff
@@ -0,0 +1,85 @@
+package com.thebois.models.inventory;
+
+import java.util.ArrayList;
+import java.util.Optional;
+
+import com.thebois.models.inventory.items.IItem;
+import com.thebois.models.inventory.items.ItemType;
+
+/**
+ * Base implementation for all inventories.
+ */
+public abstract class AbstractInventory {
+
+    private final ArrayList<IItem> items = new ArrayList<>();
+
+    /**
+     * Inputs an item to the inventory.
+     *
+     * @param item The item to be added to the inventory.
+     */
+    public void add(final IItem item) {
+        items.add(item);
+    }
+
+    /**
+     * Removes and returns the first item of the specified type.
+     *
+     * @param itemType The item type to be taken.
+     *
+     * @return The item removed from the inventory.
+     */
+    public Optional<IItem> take(final ItemType itemType) {
+
+        Optional<IItem> item = Optional.empty();
+        if (doesItemTypeExist(itemType)) {
+            final int firstIndexOfItemType = getFirstIndexOf(itemType);
+            item = Optional.ofNullable(items.remove(firstIndexOfItemType));
+        }
+        return item;
+    }
+
+    private int getFirstIndexOf(final ItemType itemType) {
+        int index = -1;
+        for (final IItem item : items) {
+            if (item.getType().equals(itemType)) {
+                index = items.indexOf(item);
+            }
+        }
+        return index;
+    }
+
+    /**
+     * Check if an item type is in the current inventory.
+     *
+     * @param itemType The item type to be checked for.
+     *
+     * @return A boolean indicating if the item is in the inventory or not.
+     */
+    private boolean doesItemTypeExist(final ItemType itemType) {
```

Comment:
style

Doesn't really matter since it's a private method, but it could just be called `itemExists(ItemType)` or `hasItem(ItemType`.

Reply:
style
I agree with Martin. `isItemPresent(ItemType)` could be another way of putting it.

Reply:
👍